### ISSUE TITLE
Migrations don't work when using TypeScript paths and module aliases

### ISSUE THREAD

**Issue by arielweinberger** [typeorm/typeorm#1556]
**Issue type:**

[ ] question
[X] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[X] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `websql`

**TypeORM version:**

[X] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**
Set up migrations and include them in your configuration file (migrationsDir).

In one of the entities, for example, there is the following import:
```ts
import { AccountEntity } from '@app/resources/entities';
```

When running the migrations, the following error shows:
```
Error: Cannot find module '@app/resources/entities'
    at Function.Module._resolveFilename (module.js:513:15)
    at Function.Module._load (module.js:463:25)
    at Module.require (module.js:556:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (D:\mypath\build\resources\entities\Account.entity.js:12:20)
    at Module._compile (module.js:612:30)
    at Object.Module._extensions..js (module.js:623:10)
    at Module.load (module.js:531:32)
    at tryModuleLoad (module.js:494:12)
```

But it works if I change the source code for the entity file to a relative path, for example:
```
import { AccountEntity } from './Account.entity';
```

Important to mention, I'm also registering the module properly using moduleAliases in package.json and tsconfig.json and they work in production.

```
  // package.json
  "_moduleAliases": {
    "@app": "./build",
    "@app/resources/entities": "./build/resources/entities",
  }

  // tsconfig.json
    "paths": {
      "@app/resources/*": ["resources/*"],
    }
```

Basically the problem is that migrations don't work with module aliases/typescript paths.

**Comment by xipheCom** [comment]
I do not know if you or typeorm use ts-node behind the scene. But ts-node knows nothing about tsconfig paths section. There is a module for it **tsconfig-paths**. Instead of `ts-node migration/xxxx.ts` you have to start it with `ts-node -r tsconfig-paths/register`. Of course, install tsconfig-paths first. If typeorm do migrations with ts-node that could be the problem. But I am knew to this.

**Comment by dtunes** [comment]
Had the same problem a few days ago, wish I found this earlier, if it wasn't for the default "is:open" flag in the issues search "I always forget to remove it".

packages.json
```json
"scripts": {
"typeorm": "ts-node -r tsconfig-paths/register ./node_modules/.bin/typeorm",
"typeorm:sync": "yarn typeorm schema:sync",
"typeorm:drop": "yarn typeorm schema:drop",
"typeorm:reset": "yarn typeorm:drop && yarn typeorm:sync"
}
```

**Comment by RaiMX** [comment]
```
"scripts": {
  ...
  "typeorm": "ts-node -r tsconfig-paths/register ./node_modules/.bin/typeorm",
  ...
}
```
then for example:
```
npm run typeorm migration:generate ./src/migrations -- -d ./src/data-source.ts
```

**Comment by ertankara** [comment]
If you are using `typeorm-ts-node-commonjs` then you can follow these steps to make it work:

1. Install module-alias
2. In your data-source add a plain import as `import "module-alias/register"`
3. In your package.json, add following field, same level as scripts, dependencies etc. 
```
  "_moduleAliases": {
    "src": "./src"
  },
```
4. Add this as your migration generation command, update according to your needs: `npx typeorm-ts-node-commonjs migration:generate ./src/migrations/$npm_config_name -d ./src/config/data-source.ts`
5. Make sure to update your data-source so that it reads migrations and entities under `src/`
6. npm run migration:generate --name=migrationName

How your data-source migration and entities fields should look like:

```
  entities: ['src/**/*.entity.ts'],
  migrations: ['src/migrations/*.ts'],
```

**Comment by khuongln-1346** [comment]
I found a simple solution [here](https://stackoverflow.com/a/77996086)

**Comment by FarahWe** [comment]
I hate typeorm

**Comment by aymather** [comment]
For anyone using a monorepo structure, you might need to point at the project-specific tsconfig.app.json like this: `npx ts-node --project apps/microservice-name/tsconfig.app.json -r tsconfig-paths/register ./node_modules/.bin/typeorm`

**Comment by Rechdan** [comment]
### For anyone having this issue

#### tsconfig.json

```json
{
  "ts-node": {
    "transpileOnly": true,
    "require": [
      "tsconfig-paths/register"
    ],
    "compilerOptions": {
      "module": "CommonJS"
    }
  },
  ...
}
```

#### cmd

```sh
typeorm-ts-node-commonjs migration:show -d ./datasource.ts
typeorm-ts-node-commonjs migration:run -d ./datasource.ts
typeorm-ts-node-commonjs migration:revert -d ./datasource.ts
```

You welcome 🤝

**Comment by MatthiasGruba** [comment]
```
{
  "ts-node": {
    "transpileOnly": true,
    "require": [
      "tsconfig-paths/register"
    ],
    "compilerOptions": {
      "module": "CommonJS"
    }
  },
  ...
}
```

@Rechdan works! Only the `"require": [...]` is really needed but the `transpileOnly` speeds it up x3!
